File: 10. Regular Expression Matching/Main.java
class Solution {
    public boolean isMatch(String s, String p) {
        StringBuilder regexPat = new StringBuilder();
        int pLength = p.length();
        for (int i = 0; i < pLength; ++i) {
            char c = p.charAt(i);
            if (i + 1 < pLength && p.charAt(i + 1) == '*') {
                if (c == '.') {
                    regexPat.append(".*");
                } else {
                    regexPat.append(c).append('*');
                }
                ++i; // skip '*'
            } else {
                regexPat.append(c);
            }
        }
        String regex = "^" + regexPat.toString() + "$";
        return s.matches(regex);
    }
}

File: 10. Regular Expression Matching/main.cpp
#include <string>
#include <regex>
using namespace std;

class Solution {
public:
    bool isMatch(string s, string p) {
        string regexPat = "";
        int pLength = p.length();
        for(int i = 0; i < pLength; ++i){
            if(i+1 < pLength && p[i + 1] == '*'){
                regexPat += p[i] == '.' ? ".*" : "(" + string(1, p[i]) + ")*";
                ++i;
            } else {
                regexPat += p[i];
            }
        }
        return regex_match(s, regex("^" + regexPat + "$"));
    }
};